src/better-shifting.cow

```diff
@@ -422,10 +422,191 @@ or by a scalar shift amount which applies to all elements in the \tcode{simd::ve
+  template<class T>
+    constexpr T rotr(T x, int s) noexcept;
+
+  \html{[\N{HORIZONTAL ELLIPSIS}]}
+}
+}
+}
+
+In \eelis{bit}, add a new subclause immediately preceding \eelis{bit.rotate}:
+
+\Bins{
+\html-h3{Shifting \stable-ref{bit.shift}}
+
+\itemdecl{
+template<class T>
+  constexpr T shl(T x, int s) noexcept;
+}
+
+\itemdescr{
+\constraints
+\tcode{T} is a signed or unsigned integer type\iref{basic.fundamental}.
+
+\effects
+If \tcode{s} is negative, the behavior is erroneous and an
+implementation-defined value is returned. Otherwise, returns the value of
+\tcode{x} multiplied by 2\sup{\tcode{s}}.
+\wg21_note{If \tcode{s} is greater than or equal to the bit width of \tcode{T},
+the result is zero.}
+}
+
+\itemdecl{
+template<class T>
+  constexpr T shr(T x, int s) noexcept;
+}
+
+\itemdescr{
+\constraints
+\tcode{T} is a signed or unsigned integer type\iref{basic.fundamental}.
+
+\effects
+If \tcode{s} is negative, the behavior is erroneous and an
+implementation-defined value is returned. Otherwise, returns the value of
+\tcode{x} divided by 2\sup{\tcode{s}}, rounded toward negative infinity.
+\wg21_note{If \tcode{s} is greater than or equal to the bit width of \tcode{T},
+the result is -1 if \tcode{x} is negative and 0 otherwise; the computation of
+2\sup{\tcode{s}} does not overflow.}
+}
+}
+
+\h3{[simd]}
+
+In \eelis{simd.syn}, change the synopsis as follows:
+
+\Bdiff{
+\itemdecl{
+\html{[\N{HORIZONTAL ELLIPSIS}]}
+
+// \serif{[simd.bit], Bit manipulation}
+
+\html{[\N{HORIZONTAL ELLIPSIS}]}
+
+\ins{
+template<\exposid{simd-type} V0, \exposid{simd-type} V1>
+  constexpr V0 shl(const V0& v, const V1& s) noexcept;
+template<\exposid{simd-type} V>
+  constexpr V  shl(const V& v, int s) noexcept;
+
+template<\exposid{simd-type} V0, \exposid{simd-type} V1>
+  constexpr V0 shr(const V0& v, const V1& s) noexcept;
+template<\exposid{simd-type} V>
+  constexpr V  shr(const V& v, int s) noexcept;
+}
+
+template<\exposid{simd-type} V0, \exposid{simd-type} V1>
+  constexpr V0 rotl(const V0& v, const V1& s) noexcept;
+template<\exposid{simd-type} V>
+  constexpr V  rotl(const V& v, int s) noexcept;
+
+template<\exposid{simd-type} V0, \exposid{simd-type} V1>
+  constexpr V0 rotr(const V0& v, const V1& s) noexcept;
+template<\exposid{simd-type} V>
+  constexpr V  rotr(const V& v, int s) noexcept;
+
+\html{[\N{HORIZONTAL ELLIPSIS}]}
+
+// See \serif{[simd.bit], Bit manipulation}
+using simd::byteswap;
+using simd::bit_ceil;
+using simd::bit_floor;
+using simd::has_single_bit;
+\ins{using simd::shl;}
+\ins{using simd::shr;}
+using simd::rotl;
+using simd::rotr;
+using simd::bit_width;
+using simd::countl_zero;
+using simd::countl_one;
+using simd::countr_zero;
+using simd::countr_one;
+using simd::popcount;
+
+\html{[\N{HORIZONTAL ELLIPSIS}]}
+}
+}
+
+In \eelis{simd.bit}, immediately preceding the first declaration of
+\tcode{rotl}, insert the following:
+
+\Bins{
+\itemdecl{
+template<\exposid{simd-type} V0, \exposid{simd-type} V1>
```

Comment:
Can't you just put these immediately below `rotl` and `rotr` and apply blanket wording to them? There is already precedent in [simd] for that, and it would make the wording quite a bit shorter.

Reply:
I don't have time to do this right now but I'll apply the rest of the changes.